File: lib/src/db/entities/datasource.entity.ts
import { IDatasource, IProject } from "@traceo/types";
import { BaseEntity, Column, Entity, JoinColumn, ManyToOne, PrimaryGeneratedColumn } from "typeorm";
import { Project } from "./project.entity";

@Entity()
export class Datasource extends BaseEntity implements IDatasource {
    @PrimaryGeneratedColumn("uuid")
    id?: string;

    @Column({ type: "varchar", unique: true })
    name: string;

    @Column({ type: "simple-json", nullable: false })
    config: string;

    @Column({ type: "varchar", nullable: false })
    type: DATASOURCE_PROVIDER;

    @ManyToOne(() => Project, {
        onUpdate: "CASCADE",
        onDelete: "CASCADE"
    })
    @JoinColumn({
        name: "project_id"
    })
    project: IProject;
}

enum DATASOURCE_PROVIDER {
    HTTP,
    MONGODB,
    POSTGRESQL,
    MYSQL,
    CLICKHOUSE
}

File: server/traceo-common/traceo-common-jpa/src/main/java/org/traceo/common/jpa/entities/DatasourceEntity.java
package org.traceo.common.jpa.entities;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.traceo.common.transport.dto.DatasourceConfigurationDto;
import org.traceo.common.transport.enums.DatasourceProvider;

@Entity
@Table(name = "traceo_datasource")
@Getter
@Setter
public class DatasourceEntity extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(unique = true, nullable = false)
    private String name;

    @Column(nullable = false)
    @JdbcTypeCode(value = SqlTypes.JSON)
    private DatasourceConfigurationDto config;

    @Column(nullable = false)
    private DatasourceProvider provider = DatasourceProvider.HTTP;

    @ManyToOne(cascade = {
            CascadeType.REMOVE,
            CascadeType.PERSIST
    })
    @JoinColumn(name = "project_id")
    private ProjectEntity project;
}
